### PR TITLE
Resolve relative paths on configuration loading

### PR DESCRIPTION
**Pull request by fabxc**
This moves the concern of resolving the files relative to the config
file into the configuration loading itself.
It also fixes #921 which did not load the cert and token files relatively.

@jimmidyson I missed the relative loading in my review, you could hardly know. Just be aware that this might break your setup.

@juliusv

### CODE REVIEW

**Comment by jimmidyson** [comment]
@fabxc Thanks - will keep an eye on it when we next upgrade our fork. We normally use absolute paths anyway - on kubernetes the relevant files (ca certs, bearer token files, mainly) are mounted at known locations.

**Comment by juliusv** [comment]
:+1:

**Comment by fabxc** [comment]
@juliusv added commit to validate existence of cert and token files via `promtool check-config`

**Comment by jimmidyson** [comment]
Nice - thanks @juliusv.

File: cmd/promtool/main.go [hunk 1]
```diff
@@ -99,6 +89,30 @@ func checkConfig(t cli.Term, filename string) ([]string, error) {
 		ruleFiles = append(ruleFiles, rfs...)
 	}
 
+	exists := func(fn string) bool {
+		// Nothing set, nothing to error on.
+		if fn == "" {
+			return true
+		}
+		_, err := os.Stat(fn)
+		return os.IsNotExist(err)
```
Comment (juliusv) [inline cmd/promtool/main.go]: This should return the opposite - and then the exists check below needs to be inverted as well.

File: cmd/promtool/main.go [hunk 2]
```diff
@@ -99,6 +89,30 @@ func checkConfig(t cli.Term, filename string) ([]string, error) {
 		ruleFiles = append(ruleFiles, rfs...)
 	}
 
+	exists := func(fn string) bool {
+		// Nothing set, nothing to error on.
+		if fn == "" {
+			return true
+		}
+		_, err := os.Stat(fn)
```
Comment (jimmidyson) [inline cmd/promtool/main.go]: Do we care about dead symlinks? `os.Lstat` might be better here.

Comment (jimmidyson) [inline cmd/promtool/main.go]: Sorry - forget me, mixing up `Stat` & `Lstat` again...

File: cmd/promtool/main.go [hunk 3]
```diff
@@ -99,6 +89,30 @@ func checkConfig(t cli.Term, filename string) ([]string, error) {
...
+		return !os.IsNotExist(err)
```
Comment (grobie) [inline cmd/promtool/main.go]: So if accessing the filesystem fails for whatever reason, we treat this as if the file exists?

Comment (fabxc) [inline cmd/promtool/main.go]: Obviously a bad idea – thanks.

**Comment by fabxc** [comment]
ping @grobie

File: cmd/promtool/main.go [hunk 4]
```diff
@@ -99,6 +89,30 @@ func checkConfig(t cli.Term, filename string) ([]string, error) {
...
+		if fn == "" {
+			return true
+		}
+		_, err := os.Stat(fn)
+		return err == nil
+	}
+
+	for _, scfg := range cfg.ScrapeConfigs {
+		if !found(scfg.BearerTokenFile) {
+			return nil, fmt.Errorf("bearer token file %q does not exist", scfg.BearerTokenFile)
```
Comment (grobie) [inline cmd/promtool/main.go]: The error message might be a bit misleading. Stat could fail for a lot of different reasons. Maybe use a more generic wording.

Comment (fabxc) [inline cmd/promtool/main.go]: Right again.

**Comment by grobie** [comment]
:+1:

**Pull request merged**